### DISCUSSION TITLE
Is the PnP dependency tree flat?

### DISCUSSION THREAD

**Discussion by loynoir** [yarnpkg/berry#4695] [Q&A] [answered]
https://www.kochan.io/nodejs/why-should-we-use-pnpm.html

> flattened dependency trees come with a bunch of issues:
> 1.  modules can access packages they don’t depend on
> 2.  the algorithm of flattening a dependency tree is pretty complex
> 3.  some of the packages have to be copied inside one project’s node_modules folder

I really like `pnpm` not-`flattened` design, but `pnp` is so awesome.

Question:
- Is `pnp` using `flattened` or not-`flattened` tree?
- Can modules access packages they don’t depend on?

**Comment by arcanis** [comment, marked as answer]
The PnP tree is flat, ghost dependency accesses are prevented.
